File: ios-native/Dimo/Sync/LendingSharingTransport.swift
import ConvexMobile
import Foundation

/// A Dimo account found by name or email.
struct LendUser: Decodable, Sendable, Equatable {
  var userId: String
  var name: String
  var email: String?
  var photoUrl: String?
  /// `none`, `connected`, `invited` or `invitedYou`.
  var relation: String
  /// Your contactId for them: the shared contact when connected, or the
  /// contact a pending invite is for.
  var contactId: String?
}

/// An invite waiting for this account to accept or decline.
struct IncomingLendInvite: Codable, Sendable, Identifiable, Equatable {
  var inviteId: String
  var inviterName: String
  var inviterEmail: String?
  var inviterPhotoUrl: String?
  /// Turns a stopped share back on.
  var reconnect: Bool
  var createdAt: Double

  var id: String { inviteId }
}

/// An invite this account sent that hasn't been accepted yet.
struct OutgoingLendInvite: Codable, Sendable, Identifiable, Equatable {
  var inviteId: String
  var contactName: String
  var contactId: String?
  var inviteeEmail: String?
  var inviteePhotoUrl: String?
  var createdAt: Double

  var id: String { inviteId }
}

struct LendConnectionSummary: Codable, Sendable, Identifiable, Equatable {
  var connectionId: String
  /// `dimo:<connectionId>`, the contactId every shared entry carries.
  var contactId: String
  var contactName: String
  /// `active` or `revoked`.
  var status: String
  var createdAt: Double
  var revokedAt: Double?
  /// The other member's profile photo.
  var photoUrl: String?

  var id: String { connectionId }
  var isActive: Bool { status == "active" }
}

struct SentLendInvite: Decodable, Sendable {
  var inviteId: String
}

struct AcceptedLendInvite: Decodable, Sendable {
  var connectionId: String
  var contactId: String
}

/// Whose past entries are kept when both sides already tracked each other;
/// the other side's duplicates are deleted. Clients always keep both.
enum LendHistoryChoice: String, CaseIterable, Identifiable, Sendable {
  case both
  case inviter
  case accepter

  var id: String { rawValue }
}

protocol LendingSharingAPI: Sendable {
  func searchUsers(query: String) async throws -> [LendUser]
  func reshare(connectionId: String) async throws
  func setProfilePhoto(_ photoUrl: String?) async throws
  func sendInvite(userId: String, contactId: String, contactName: String) async throws
  func accept(inviteId: String, contactId: String?, contactName: String?, history: LendHistoryChoice) async throws -> AcceptedLendInvite
  func decline(inviteId: String) async throws
  func cancel(inviteId: String) async throws
  func revoke(connectionId: String) async throws
  func incomingInvites() async throws -> [IncomingLendInvite]
  func outgoingInvites() async throws -> [OutgoingLendInvite]
  func connections() async throws -> [LendConnectionSummary]
}

/// Authenticated calls for collaborative lending (`convex/lending.ts`).
final class LendingSharingTransport: LendingSharingAPI, @unchecked Sendable {
  private let client: ConvexClientWithAuth<WorkOSSession>

  init(client: ConvexClientWithAuth<WorkOSSession>) {
    self.client = client
  }

  func searchUsers(query: String) async throws -> [LendUser] {
    try await firstValue(
      client.subscribe(to: "lending:searchLendUsers", with: ["query": query])
    )
  }

  func reshare(connectionId: String) async throws {
    let _: SentLendInvite = try await withTimeout(seconds: 45) {
      try await self.client.mutation(
        "lending:reshareLendConnection",
        with: ["connectionId": connectionId]
      )
    }
  }

  func setProfilePhoto(_ photoUrl: String?) async throws {
    let args: [String: ConvexEncodable?] = ["photoUrl": photoUrl]
    try await withTimeout(seconds: 45) {
      try await self.client.mutation("lending:setProfilePhoto", with: args)
    }
  }

  func sendInvite(userId: String, contactId: String, contactName: String) async throws {
    let _: SentLendInvite = try await withTimeout(seconds: 45) {
      try await self.client.mutation(
        "lending:sendLendInvite",
        with: ["userId": userId, "contactId": contactId, "contactName": contactName]
      )
    }
  }

  func accept(
    inviteId: String,
    contactId: String?,
    contactName: String?,
    history: LendHistoryChoice
  ) async throws -> AcceptedLendInvite {
    var args: [String: ConvexEncodable?] = ["inviteId": inviteId, "history": history.rawValue]
    if let contactId { args["contactId"] = contactId }
    if let contactName, !contactName.isEmpty { args["contactName"] = contactName }
    let sendable = args
    return try await withTimeout(seconds: 45) {
      try await self.client.mutation("lending:acceptLendInvite", with: sendable)
    }
  }

  func decline(inviteId: String) async throws {
    try await withTimeout(seconds: 45) {
      try await self.client.mutation("lending:declineLendInvite", with: ["inviteId": inviteId])
    }
  }

  func cancel(inviteId: String) async throws {
    try await withTimeout(seconds: 45) {
      try await self.client.mutation("lending:cancelLendInvite", with: ["inviteId": inviteId])
    }
  }

  func revoke(connectionId: String) async throws {
    try await withTimeout(seconds: 45) {
      try await self.client.mutation(
        "lending:revokeLendConnection",
        with: ["connectionId": connectionId]
      )
    }
  }

  func incomingInvites() async throws -> [IncomingLendInvite] {
    try await firstValue(
      client.subscribe(to: "lending:listIncomingLendInvites", with: [:] as [String: ConvexEncodable?])
    )
  }

  func outgoingInvites() async throws -> [OutgoingLendInvite] {
    try await firstValue(
      client.subscribe(to: "lending:listOutgoingLendInvites", with: [:] as [String: ConvexEncodable?])
    )
  }

  func connections() async throws -> [LendConnectionSummary] {
    try await firstValue(
      client.subscribe(to: "lending:listLendConnections", with: [:] as [String: ConvexEncodable?])
    )
  }
}

File: android-native/app/src/main/java/app/dimo/android/sync/LendingSharingTransport.kt
package app.dimo.android.sync

import app.dimo.android.auth.WorkOSSession
import dev.convex.android.ConvexClientWithAuth
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull

/**
 * Collaborative lending calls (`convex/lending.ts`). Port of
 * `ios-native/Dimo/Sync/LendingSharingTransport.swift`.
 */
/** A Dimo account found by its verified email. */
@Serializable
data class LendUser(
  val userId: String,
  val name: String,
  val email: String,
  /** `none`, `self`, `connected`, `invited` or `invitedYou`. */
  val relation: String,
  /**
   * Your contactId for them: the shared ledger when connected, or the contact
   * a pending invite is for.
   */
  val contactId: String? = null,
)

/** An invite waiting for this account to accept or decline. */
@Serializable
data class IncomingLendInvite(
  val inviteId: String,
  val inviterName: String,
  val inviterEmail: String? = null,
  val createdAt: Double,
)

/** An invite this account sent that hasn't been accepted yet. */
@Serializable
data class OutgoingLendInvite(
  val inviteId: String,
  val contactName: String,
  val contactId: String? = null,
  val inviteeEmail: String? = null,
  val createdAt: Double,
)

@Serializable
data class LendConnectionSummary(
  val connectionId: String,
  /** `dimo:<connectionId>`, the contactId every shared entry carries. */
  val contactId: String,
  val contactName: String,
  /** `active` or `revoked`. */
  val status: String,
  val createdAt: Double,
  val revokedAt: Double? = null,
) {
  val isActive: Boolean get() = status == "active"
}

@Serializable
data class AcceptedLendInvite(val connectionId: String, val contactId: String)

@Serializable
data class VerifiedEmailResult(
  /** False when the deployment has no WorkOS API key configured. */
  val available: Boolean,
  val email: String? = null,
)

/**
 * Whose past entries make up the shared ledger when both sides already tracked
 * each other; the other side's duplicates are deleted.
 */
enum class LendHistoryChoice(val wire: String) {
  BOTH("both"),
  INVITER("inviter"),
  ACCEPTER("accepter"),
}

class LendingSharingTransport(
  private val client: ConvexClientWithAuth<WorkOSSession>,
) {
  private val json = Json { ignoreUnknownKeys = true }

  suspend fun findUser(email: String): LendUser? {
    val element = query("lending:findLendUser", mapOf("email" to email))
    if (element == null || element is JsonNull) return null
    return decode(LendUser.serializer(), element)
  }

  suspend fun sendInvite(userId: String, contactId: String, contactName: String) {
    mutation(
      "lending:sendLendInvite",
      mapOf("userId" to userId, "contactId" to contactId, "contactName" to contactName),
    )
  }

  suspend fun accept(
    inviteId: String,
    contactId: String?,
    contactName: String?,
    history: LendHistoryChoice,
  ): AcceptedLendInvite {
    val args = buildMap<String, Any?> {
      put("inviteId", inviteId)
      put("history", history.wire)
      if (contactId != null) put("contactId", contactId)
      if (!contactName.isNullOrEmpty()) put("contactName", contactName)
    }
    return decode(AcceptedLendInvite.serializer(), mutation("lending:acceptLendInvite", args))
  }

  suspend fun decline(inviteId: String) {
    mutation("lending:declineLendInvite", mapOf("inviteId" to inviteId))
  }

  suspend fun cancel(inviteId: String) {
    mutation("lending:cancelLendInvite", mapOf("inviteId" to inviteId))
  }

  suspend fun revoke(connectionId: String) {
    mutation("lending:revokeLendConnection", mapOf("connectionId" to connectionId))
  }

  suspend fun incomingInvites(): List<IncomingLendInvite> =
    decodeList(IncomingLendInvite.serializer(), query("lending:listIncomingLendInvites", emptyMap()))

  suspend fun outgoingInvites(): List<OutgoingLendInvite> =
    decodeList(OutgoingLendInvite.serializer(), query("lending:listOutgoingLendInvites", emptyMap()))

  suspend fun connections(): List<LendConnectionSummary> =
    decodeList(LendConnectionSummary.serializer(), query("lending:listLendConnections", emptyMap()))

  suspend fun refreshVerifiedEmail(): VerifiedEmailResult {
    val element = withTimeout(TIMEOUT_MS) {
      client.action<JsonElement>("lendingEmail:refreshVerifiedEmail", emptyMap())
    }
    return decode(VerifiedEmailResult.serializer(), element)
  }

  private suspend fun mutation(name: String, args: Map<String, Any?>): JsonElement =
    withTimeout(TIMEOUT_MS) { client.mutation<JsonElement>(name, args) }

  private suspend fun query(name: String, args: Map<String, Any?>): JsonElement? =
    withTimeout(TIMEOUT_MS) { client.subscribe<JsonElement>(name, args).first().getOrThrow() }

  private fun <T> decode(serializer: KSerializer<T>, element: JsonElement?): T =
    json.decodeFromJsonElement(serializer, element ?: JsonNull)

  private fun <T> decodeList(serializer: KSerializer<T>, element: JsonElement?): List<T> =
    if (element == null || element is JsonNull) {
      emptyList()
    } else {
      json.decodeFromJsonElement(ListSerializer(serializer), element)
    }

  private companion object {
    const val TIMEOUT_MS = 45_000L
  }
}
